src/marinade.ts

```diff
@@ -743,4 +752,244 @@ export class Marinade {
       transaction,
     }
   }
+
+  /**
+   * Returns a transaction with the instructions to
+   * Deposit an amount of stake pool tokens.
+   *
+   * @param {web3.PublicKey} stakePoolTokenAddress - The stake pool token account to be deposited
+   * @param {number} amountToDeposit - Amount to deposit
+   * @param {ValidatorStats[]} validators - List of validators to prio where to take the stake from
+   */
+  async depositStakePoolToken(
+    stakePoolTokenAddress: web3.PublicKey,
+    amountToDeposit: number,
+    validators: ValidatorStats[]
+  ): Promise<MarinadeResult.LiquidateStakePoolToken> {
+    const marinadeState = await this.getMarinadeState()
+    const ownerAddress = assertNotNullAndReturn(
+      this.config.publicKey,
+      ErrorMessage.NO_PUBLIC_KEY
+    )
+
+    const lookupTable = (
+      await this.config.connection.getAddressLookupTable(
+        this.config.lookupTableAddress
+      )
+    ).value
+    if (!lookupTable) {
+      throw new Error('Failed to load the lookup table')
+    }
+
+    const expectedSOL = await computeExpectedSOL(
+      amountToDeposit,
+      this.config.connection,
+      stakePoolTokenAddress
+    )
+
+    // Due to our contract 1 SOL limit for Stake accounts we can't accept less than equivalent of 1 SOL
+    if (expectedSOL / LAMPORTS_PER_SOL < 1) {
+      throw new Error("Can't convert less than equivalent of 1 SOL")
+    }
+
+    const { blockhash: recentBlockhash } =
+      await this.config.connection.getLatestBlockhash('finalized')
+
+    const instructions: web3.TransactionInstruction[] = []
+
+    const validatorSet = new Set(
+      validators.filter(v => v.score).map(v => v.vote_account)
+    )
+    const withdrawTx = await withdrawStake(
+      this.provider.connection,
+      stakePoolTokenAddress,
+      ownerAddress,
+      amountToDeposit,
+      undefined,
+      undefined,
+      undefined,
+      undefined,
+      (a: any, b: any) => selectSpecificValidator(a, b, validatorSet)
+    )
+
+    instructions.push(...withdrawTx.instructions)
+
+    const {
+      associatedTokenAccountAddress: associatedMSolTokenAccountAddress,
+      createAssociateTokenInstruction,
+    } = await getOrCreateAssociatedTokenAccount(
+      this.provider,
+      marinadeState.mSolMintAddress,
+      ownerAddress
+    )
+
+    if (createAssociateTokenInstruction) {
+      instructions.push(createAssociateTokenInstruction)
+    }
+
+    const { duplicationFlag, validatorIndex } = await identifyValidatorFromTx(
+      withdrawTx.instructions,
+      this.provider,
+      marinadeState
+    )
+
+    const depositInstruction =
+      await this.marinadeFinanceProgram.depositStakeAccountInstructionBuilder({
```

Comment:
Without going super deep into the code here. In my understanding we are not using the referral program at all without this line?
```const program = this.provideReferralOrMainProgram()```